### PR TITLE
[dockerfile]: improvements

### PR DESCRIPTION
**Pull request by alebabai**
## what
- use multi-stage build to separate packages compilation from production image
- more cacheble layers structure
- group common envs into single instruction
- install compiled libs to bastion image

## why
#27

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,124 +1,146 @@
...
+    rm -rf duo_unix
+
+## Install Google Authenticator PAM module
+COPY --from=google-authenticator-libpam-builder google-authenticator-libpam google-authenticator-libpam
+RUN (cd google-authenticator-libpam && make install) && \
+    rm -rf google-authenticator-libpam
+
+## Install OpenSSH Portable
+COPY --from=openssh-portable-builder openssh-portable openssh-portable
+RUN (cd openssh-portable && make install) && \
```
Comment (osterman) [inline Dockerfile]: What about installing in the builder image to a prefix like `/dist`, then doing a `COPY --from=openssh-portable-builder /dist /`

File: Dockerfile [hunk 2]
```diff
@@ -1,124 +1,146 @@
...
-ENV SSH_AUTHORIZED_KEYS_COMMAND=none
-ENV SSH_AUTHORIZED_KEYS_COMMAND_USER=nobody
+## Install Duo
+COPY --from=duo-builder duo_unix duo_unix
+RUN (cd duo_unix && make install) && \
+    rm -rf duo_unix
+
+## Install Google Authenticator PAM module
+COPY --from=google-authenticator-libpam-builder google-authenticator-libpam google-authenticator-libpam
+RUN (cd google-authenticator-libpam && make install) && \
```
Comment (osterman) [inline Dockerfile]: See https://github.com/cloudposse/bastion/pull/29/files#r219734929

File: Dockerfile [hunk 3]
```diff
@@ -1,124 +1,146 @@
...
+RUN chmod 755 /usr/bin/sudosh
 
-#
-# SSH
-#
-ENV SSH_AUTHORIZED_KEYS_COMMAND=none
-ENV SSH_AUTHORIZED_KEYS_COMMAND_USER=nobody
+## Install Duo
+COPY --from=duo-builder duo_unix duo_unix
+RUN (cd duo_unix && make install) && \
```
Comment (osterman) [inline Dockerfile]: See https://github.com/cloudposse/bastion/pull/29/files#r219734929

File: Dockerfile (hunk 2 above)
Comment (osterman) [reply Dockerfile]: Are the `(....)` necessary? Logically don't seem like it.

File: Dockerfile (hunk 1 above)
Comment (osterman) [reply Dockerfile]: Are the `(....)` necessary? Logically don't seem like it.

Comment (alebabai) [reply Dockerfile]: Agree with you, refactored code to use common approach with `dist` folder

**Pull request merged**